lib/pbench/cli/agent/commands/tools/kill.py

```diff
@@ -0,0 +1,174 @@
+
+from pbench.cli.agent import pass_cli_context
+from pbench.cli.agent.base import BaseCommand
+from pbench.cli.agent.options import common_options
+
+
+class KillTools(BaseCommand):
+    """Find and stop all orchestrated Tool Meister instances."""
+
+    def __init__(self):
+        super().__init__()
+        self.uuid_to_rundir = {}
+
+    def gen_run_directories(self) -> Tuple(pathlib.Path, str):
+        """Generate the list of run directories available under ${pbench_run},
+        yielding a Path object for that directory, along with its recorded
+        UUID.
+
+        Yields a tuple of the run directory Path object, and associated UUID.
+        """
+        for entry in self.pbench_run.iterdir():
+            if not entry.is_dir():
+                continue
+            tm_dir = entry / "tm"
+            try:
+                uuid = tm_dir / ".uuid".read_text()
+            except FileNotFoundError:
+                # This is either not a pbench run directory, or the Tool Meister
+                # sub-system was not orchestrated by pbench-tool-meister-start
+                # for this run.
+                continue
+            yield tm_dir, uuid
+
+    def gen_local_pids(self) -> Tuple(str, psutil.Process):
+        """Generate all the local PIDs from all run directories.
+
+        Yields a tuple of the UUID, psutil.Process pair derived from the PID
+        value read from the file.
+
+
+        All the Redis server PIDs are yielded first, then the Tool Data Sinks,
+        and finally the local Tool Meisters.
+        """
+        tds_pid_file_name = "pbench-tool-data-sink.pid"
+        tm_pid_file_name = "tm.pid"
+        redis_pid_file_name = "redis.pid"
+
+        tds_pids = {}
+        tm_pids = {}
+        redis_pids = {}
+
+        for tm_dir, uuid in self.gen_run_directories():
+            # If a run directory has any dangling components of the Tool
+            # Meister sub-system active, that is PID files for any local
+            # component, record those components.
+            try:
+                tds_pid = (tm_dir / tds_pid_file_name).read_text()
+            except FileNotFoundError:
+                pass
+            else:
+                tds_pids[uuid] = psutil.Process(tds_pid)
+
+            try:
+                tm_pid = (tm_dir / tm_pid_file_name).read_text()
+            except FileNotFoundError:
+                pass
+            else:
+                tm_pids[uuid] = psutil.Process(tm_pid)
+
+            try:
+                redis_pid = (tm_dir / redis_pid_file_name).read_text()
+            except FileNotFoundError:
+                pass
+            else:
+                redis_pids[uuid] = psutil.Process(redis_pid)
+
+            if any(uuid in _pids for _pids in [tds_pids, tm_pids, redis_pids]):
+                self.uuid_to_rundir[uuid] = tm_dir
+
+        # Kill all the Redis Servers.  We do this first in case this causes
+        # all the other processes to just exit on their own.
```

Comment:
Maybe reword the comments to describe what we're doing here rather than what we expect the consumer to do with the values we yield?